File: src/network/client/TCPListener.java
package network.client;

import common.OsUtils;
import main.PushLocal;

import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * Created by brian on 8/30/15.
 */
public class TCPListener extends Thread {
    public static final int PACKET_SIZE = 1024;
    private Socket socket;
    private NetClient netClient;
    private ConcurrentLinkedQueue<String> logs;
    private InetAddress address;

    public TCPListener(NetClient netClient, ConcurrentLinkedQueue<String> logs, InetAddress address) {
        this.netClient = netClient;
        this.logs = logs;
        this.address = address;
    }

    @Override
    public void run() {
        super.run();

        try {
            socket = new Socket(address, 5577);
            sendMessage("connected");

            byte[] data = new byte[PACKET_SIZE];

            while (netClient.isRunning() && !socket.isClosed()) {
                int len = socket.getInputStream().read(data);
                if (len < 0)
                    break;
                logs.add("Received Message from " + socket.getInetAddress().getHostName());
                handleMessage(data);
                data = new byte[PACKET_SIZE];
                Thread.sleep(0);
            }
        } catch (IOException | InterruptedException e) {
            logs.add(e.getMessage());
        }

        if (socket.isConnected()) {
            try {
                dispose();
            } catch (IOException e) {
                logs.add(e.getMessage());
            }
        }

        logs.add("TCP socket disconnected");
    }

    private void handleMessage(byte[] data) throws IOException {
        // TODO: Detemine what file type was sent
        String msg = new String(data);
        msg = msg.trim();
        logs.add("TCP Message recieved: " + msg);

        if (msg.contains("notification")) {
            String[] split = msg.split(NetClient.UNIT);
            String from = split[1]; // split[0] is the type of message, in this case, "notification"
            String title = split[2];
            String text = split[3];
            String subText = split[4];
            logs.add("From: " + from + "Title: " + title + " Text: " + text + " Subtext: " + subText);
            PushLocal.fetch().postNotification(from,
                    title, text, subText);
        }
        else if (msg.contains("Indeed,")) {
            sendMessage("Nice comma.");
        }
    }

    private void sendMessage(String msg) {
        try {
            socket.getOutputStream().write(msg.getBytes());
            logs.add("Sent TCP Message: " + msg);
        } catch (IOException e) {
            logs.add(e.getMessage());
        }
    }

    public void dispose() throws IOException {
        socket.close();
    }
}

File: src/network/client/TCPListener.kt
package network.client

import main.PushLocal

import java.io.IOException
import java.net.InetAddress
import java.net.Socket
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * Created by brian on 8/30/15.
 */
class TCPListener(private val netClient: NetClient, private val logs: ConcurrentLinkedQueue<String>, private val address: InetAddress) : Thread() {
    private var socket: Socket? = null

    override fun run() {
        super.run()

        try {
            socket = Socket(address, 5577)
            sendMessage("connected")

            var data = ByteArray(PACKET_SIZE)

            while (netClient.isRunning && !socket!!.isClosed) {
                val len = socket!!.inputStream.read(data)
                if (len < 0)
                    break
                logs.add("Received Message from " + socket!!.inetAddress.hostName)
                handleMessage(data)
                data = ByteArray(PACKET_SIZE)
                Thread.sleep(0)
            }
        } catch (e: IOException) {
            logs.add(e.message)
        } catch (e: InterruptedException) {
            logs.add(e.message)
        }

        if (socket!!.isConnected) {
            try {
                dispose()
            } catch (e: IOException) {
                logs.add(e.message)
            }

        }

        logs.add("TCP socket disconnected")
    }

    @Throws(IOException::class)
    private fun handleMessage(data: ByteArray) {
        // TODO: Detemine what file type was sent
        var msg = String(data)
        msg = msg.trim { it <= ' ' }
        logs.add("TCP Message recieved: " + msg)

        if (msg.contains("notification")) {
            val split = msg.split(NetClient.UNIT.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val from = split[1] // split[0] is the type of message, in this case, "notification"
            val title = split[2]
            val text = split[3]
            val subText = split[4]
            logs.add("From: " + from + "Title: " + title + " Text: " + text + " Subtext: " + subText)
            PushLocal.singleton?.postNotification(from,
                    title, text, subText)
        } else if (msg.contains("Indeed,")) {
            sendMessage("Nice comma.")
        }
    }

    private fun sendMessage(msg: String) {
        try {
            socket!!.outputStream.write(msg.toByteArray())
            logs.add("Sent TCP Message: " + msg)
        } catch (e: IOException) {
            logs.add(e.message)
        }

    }

    @Throws(IOException::class)
    fun dispose() {
        socket?.close()
    }

    companion object {
        val PACKET_SIZE = 1024
    }
}

